### PR TITLE
feat: Add reset() method, gRPC keep-alive, and transport resilience

### PR DESCRIPTION
**Pull request by lukekim**
## Changes

- Add public `reset()` method to `SpiceClient` for recovering from unrecoverable transport failures (SSL cert mismatch, persistent UNAVAILABLE errors, stale connections pinned to wrong backend IPs)
- Extract `buildFlightClient()` for reuse during construction and after `reset()`
- Configure HTTP/2 keep-alive (30s interval, 10s timeout) on the gRPC channel to detect dead/idle connections behind load balancers
- Add `ensureFlightClient()` guard in `queryInternal()` for safety
- Handle null `flightClient` in `close()` after reset
- Add `ResetTest.java` with 20 unit tests covering happy path, edge cases, concurrency, construction variants, and integration
- Document transport resilience and `reset()` usage in README

### CODE REVIEW

File: src/test/java/ai/spice/ResetTest.java [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+        for (int i = 0; i < 3; i++) {
+            client.reset();
+            try {
+                FlightStream stream = client.query("SELECT 1");
+                stream.close();
+            } catch (Exception e) {
+                // Connection errors are fine — we're testing the reset/rebuild cycle,
+                // not server availability. NPE would indicate a broken rebuild.
+                assertFalse("Cycle " + i + ": NPE after reset means rebuild is broken",
+                        e instanceof NullPointerException);
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java:196]: Addressed in d8c2328.

File: src/test/java/ai/spice/ResetTest.java [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+package ai.spice;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.apache.arrow.flight.FlightStream;
+import org.apache.arrow.vector.VectorSchemaRoot;
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java]: Addressed in d8c2328.

File: src/test/java/ai/spice/ResetTest.java [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+            client.reset();
+            // auto-close on scope exit
+        }
+    }
+
+    // ==================== Concurrent reset() ====================
+
+    /**
+     * Concurrent calls to reset() from multiple threads should not throw
+     * or corrupt state (all methods are synchronized).
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java]: Addressed in d8c2328.

File: src/test/java/ai/spice/ResetTest.java [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+    }
+
+    /**
+     * After reset(), close() should complete without errors
+     * (the Flight client is already nulled out; close handles null gracefully).
+     */
+    public void testCloseAfterReset() throws Exception {
+        SpiceClient client = SpiceClient.builder().build();
+        client.reset();
+        client.close(); // should not throw
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java:113]: Addressed in d8c2328.

File: src/main/java/ai/spice/SpiceClient.java [hunk 5]
```diff
@@ -153,6 +153,7 @@ public class SpiceClient implements AutoCloseable {
     private FlightSqlClient flightClient;
     private CredentialCallOption authCallOptions = null;
     private BufferAllocator allocator;
+    private boolean closed = false;
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/main/java/ai/spice/SpiceClient.java]: Addressed in d8c2328.

File: src/main/java/ai/spice/SpiceClient.java [hunk 6]
```diff
@@ -282,13 +306,72 @@ public SpiceClient(String appId, String apiKey, URI flightAddress, URI httpAddre
         client.handshake(new CredentialCallOption(new BasicAuthCredentialWriter(this.appId, this.apiKey)));
         this.authCallOptions = authFactory.getCredentialCallOption();
         this.flightClient = new FlightSqlClient(client);
-        
-        // Initialize cached retryers (immutable, built once)
-        initRetryers();
-        
-        logger.debug("SpiceClient initialized (authenticated) - flightAddress={}, appId={}, target={}", this.flightAddress, this.appId, target);
+
+        logger.debug("Flight client built (authenticated) - target={}, appId={}", target, this.appId);
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/main/java/ai/spice/SpiceClient.java]: Addressed in d8c2328.

File: src/main/java/ai/spice/SpiceClient.java [hunk 7]
```diff
@@ -930,6 +1014,10 @@ private boolean shouldRetry(CallStatus status) {
 
     @Override
     public void close() throws Exception {
+        if (closed) {
+            return;
+        }
+        closed = true;
         logger.debug("Closing SpiceClient");
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/main/java/ai/spice/SpiceClient.java]: Addressed in d8c2328.

File: src/test/java/ai/spice/ResetTest.java [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+                }
+            } catch (InterruptedException ignored) {
+            } finally {
+                doneLatch.countDown();
+            }
+        }).start();
+
+        startLatch.countDown();
+        doneLatch.await();
+
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java]: Addressed in d8c2328.

File: src/test/java/ai/spice/ResetTest.java [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+                .withMaxRetries(5)
+                .withArrowMemoryLimitMB(256)
+                .withUserAgent("TestApp/1.0")
+                .build();
+        client.reset();
+
+        try {
+            FlightStream stream = client.query("SELECT 1");
+            stream.close();
+        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java:414]: Addressed in d8c2328.

**Review by sgrebnov — APPROVED** [review]

File: .github/workflows/build.yaml [hunk 10]
```diff
@@ -178,3 +186,40 @@
...
+      - name: OWASP Dependency-Check
+        run: mvn dependency-check:check -B
+
+      - name: Upload dependency-check report
+        if: always()
+        uses: actions/upload-artifact@v4
+        with:
+          name: dependency-check-report
+          path: target/dependency-check-report.html
+          retention-days: 30
```
[the remark this answers is not in the document]
Comment (lukekim) [reply .github/workflows/build.yaml:237]: Addressed in 55ea00e — added `permissions: {contents: read}` to the quality job.

File: src/test/java/ai/spice/ResetTest.java [hunk 11]
```diff
@@ -0,0 +1,463 @@
...
+     */
+    public void testQueryAfterResetRebuildsClient() throws Exception {
+        SpiceClient client = SpiceClient.builder().build();
+        client.reset();
+
+        try {
+            FlightStream stream = client.query("SELECT 1");
+            // If a local Spice runtime is running, this succeeds
+            stream.next();
+            stream.close();
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java]: Fixed — wrapped FlightStream in try-with-resources in 947bb51.

File: src/test/java/ai/spice/ResetTest.java [hunk 12]
```diff
@@ -0,0 +1,463 @@
...
+    public void testQueryWithParamsAfterResetRebuildsClient() throws Exception {
+        SpiceClient client = SpiceClient.builder().build();
+        client.reset();
+
+        try {
+            ArrowReader reader = client.queryWithParams("SELECT $1", 42);
+            while (reader.loadNextBatch()) {
+                // consume
+            }
+            reader.close();
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java]: Fixed in 947bb51.

File: src/test/java/ai/spice/ResetTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java]: Fixed — wrapped FlightStream in try-with-resources in 947bb51.

File: src/test/java/ai/spice/ResetTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java]: Fixed — wrapped ArrowReader in try-with-resources in 947bb51.

File: src/test/java/ai/spice/ResetTest.java [hunk 13]
```diff
@@ -0,0 +1,463 @@
...
+            client.reset();
+
+            // Second query (lazy rebuild)
+            FlightStream stream2 = client.query(
+                    "SELECT c_custkey FROM tpch.customer LIMIT 2");
+            int rows2 = 0;
+            while (stream2.next()) {
+                rows2 += stream2.getRoot().getRowCount();
+            }
+            assertEquals("Second query after reset should return 2 rows", 2, rows2);
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java]: Fixed — both streams wrapped in try-with-resources in 947bb51.

File: src/main/java/ai/spice/SpiceClient.java [hunk 14]
```diff
@@ -907,6 +1001,7 @@ public void refreshDataset(String dataset, RefreshOptions refreshOptions) throws
     }
 
     private FlightStream queryInternal(String sql) {
+        ensureFlightClient();
         FlightInfo flightInfo = this.flightClient.execute(sql, authCallOptions);
         Ticket ticket = flightInfo.getEndpoints().get(0).getTicket();
         return this.flightClient.getStream(ticket, authCallOptions);
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/main/java/ai/spice/SpiceClient.java]: Already fixed — queryInternal() uses snapshot-under-lock pattern. ensureFlightClient() now also has a closed guard (947bb51).

File: src/main/java/ai/spice/SpiceClient.java [hunk 15]
```diff
@@ -246,49 +267,122 @@ public SpiceClient(String appId, String apiKey, URI flightAddress, URI httpAddre
...
-        if (Strings.isNullOrEmpty(userAgent)) {
-            uaString = Config.getUserAgent();
-        } else {
-            // Prepend the user-supplied user agent string with the Spice.ai user agent
-            uaString = userAgent + " " + Config.getUserAgent();
+    /**
+     * Ensures the Flight client is connected, rebuilding it if necessary
+     * (e.g. after a {@link #reset()} call).
+     */
+    private synchronized void ensureFlightClient() {
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/main/java/ai/spice/SpiceClient.java:344]: Fixed — added closed guard in ensureFlightClient() in 947bb51.

File: src/test/java/ai/spice/ResetTest.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java]: Fixed — both streams wrapped in try-with-resources in 947bb51.

File: spotbugs-exclude.xml [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+    </Match>
+
+    <!-- RefreshOptions: public fields are part of the public API -->
+    <Match>
+        <Class name="ai.spice.RefreshOptions"/>
+        <Bug pattern="PA_PUBLIC_PRIMITIVE_ATTRIBUTE"/>
+    </Match>
+
+    <!-- Constructor-throw warnings: these constructors intentionally validate/fail early -->
+    <Match>
```
[the remark this answers is not in the document]
Comment (lukekim) [reply spotbugs-exclude.xml:39]: Fixed — narrowed to SpiceClient and SpiceClientBuilder specifically in 947bb51.

File: README.md [hunk 17]
```diff
@@ -213,6 +213,38 @@ SpiceClient client = SpiceClient.builder()
...
+    FlightStream stream = client.query(sql);
+    // process results...
+} catch (ExecutionException e) {
+    if (isTransportFailure(e.getCause())) {
+        client.reset();                     // discard bad transport, reconnect immediately
+        FlightStream stream = client.query(sql);  // no connection overhead
+    } else {
+        throw e;
+    }
+}
```
[the remark this answers is not in the document]
Comment (lukekim) [reply README.md:245]: Fixed — README example now uses try-with-resources for FlightStream in 947bb51.

File: src/test/java/ai/spice/ResetTest.java [hunk 18]
```diff
@@ -0,0 +1,463 @@
...
+        try (SpiceClient client = SpiceClient.builder().build()) {
+            // First query (establishes connection)
+            FlightStream stream1 = client.query(
+                    "SELECT c_custkey FROM tpch.customer LIMIT 1");
+            int rows1 = 0;
+            while (stream1.next()) {
+                rows1 += stream1.getRoot().getRowCount();
+            }
+            assertEquals("First query should return 1 row", 1, rows1);
+
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java]: Fixed — stream1 wrapped in try-with-resources in 947bb51.

File: src/main/java/ai/spice/SpiceClient.java [hunk 19]
```diff
@@ -246,49 +275,122 @@ public SpiceClient(String appId, String apiKey, URI flightAddress, URI httpAddre
...
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/main/java/ai/spice/SpiceClient.java:344]: Fixed — added closed guard in ensureFlightClient() in 947bb51.

File: src/main/java/ai/spice/SpiceClient.java [hunk 20]
```diff
@@ -451,14 +557,17 @@ private synchronized void initADBCIfNeeded() throws AdbcException {
         FlightSqlDriver driver = new FlightSqlDriver(allocator);
         adbcDatabase = driver.open(options);
         adbcConnection = adbcDatabase.connect();
+        adbcInitialized = true;
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/main/java/ai/spice/SpiceClient.java:589]: Fixed — now uses local temporaries; only assigns to fields after both open+connect succeed in 947bb51.

File: src/test/java/ai/spice/ResetTest.java [hunk 21]
```diff
@@ -0,0 +1,463 @@
...
+                    errors.incrementAndGet();
+                } finally {
+                    doneLatch.countDown();
+                }
+            }).start();
+        }
+
+        startLatch.countDown(); // release all threads
+        doneLatch.await();
+
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java]: Fixed — added 30s bounded timeout on all latch awaits in 947bb51.

File: spotbugs-exclude.xml [hunk 22]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  SpotBugs exclusion filter for pre-existing findings.
+  These are findings in code that predates the introduction of SpotBugs.
+  They should be addressed in follow-up PRs.
+-->
+<FindBugsFilter>
+    <!-- Example code: dead stores are intentional (demonstrating all param types) -->
+    <Match>
+        <Package name="ai.spice.example"/>
```
[the remark this answers is not in the document]
Comment (lukekim) [reply spotbugs-exclude.xml:10]: Fixed — narrowed to DLS_DEAD_LOCAL_STORE only in 947bb51.

File: spotbugs-exclude.xml [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+    <!-- RefreshOptions: public fields are part of the public API -->
+    <Match>
+        <Class name="ai.spice.RefreshOptions"/>
+        <Bug pattern="PA_PUBLIC_PRIMITIVE_ATTRIBUTE"/>
+    </Match>
+
+    <!-- Constructor-throw warnings: these constructors intentionally validate/fail early -->
+    <Match>
+        <Bug pattern="CT_CONSTRUCTOR_THROW"/>
+    </Match>
```
[the remark this answers is not in the document]
Comment (lukekim) [reply spotbugs-exclude.xml:46]: Fixed — scoped to SpiceClient and SpiceClientBuilder in 947bb51.

File: .github/workflows/build.yaml [hunk 24]
```diff
@@ -178,3 +186,42 @@ jobs:
...
+        run: mvn validate -B
+
+      - name: SpotBugs
+        run: mvn spotbugs:check -B
+
+      - name: Checkstyle
+        run: mvn checkstyle:check -B
+
+      - name: OWASP Dependency-Check
+        run: mvn dependency-check:check -B
```
[the remark this answers is not in the document]
Comment (lukekim) [reply .github/workflows/build.yaml:229]: Fixed — added dependency-check data caching and NVD_API_KEY support in 947bb51.

File: src/main/java/ai/spice/SpiceClient.java [hunk 25]
```diff
@@ -451,14 +561,17 @@ private synchronized void initADBCIfNeeded() throws AdbcException {
...
         
         logger.debug("ADBC connection established - uri={}", uri);
+        }
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/main/java/ai/spice/SpiceClient.java]: Addressed in 947bb51.

File: spotbugs-exclude.xml [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+  SpotBugs exclusion filter for pre-existing findings.
+  These are findings in code that predates the introduction of SpotBugs.
+  They should be addressed in follow-up PRs.
+-->
+<FindBugsFilter>
+    <!-- Example code: dead stores are intentional (demonstrating all param types) -->
+    <Match>
+        <Package name="ai.spice.example"/>
+    </Match>
+
```
[the remark this answers is not in the document]
Comment (lukekim) [reply spotbugs-exclude.xml:13]: Addressed in 947bb51.

File: src/test/java/ai/spice/ResetTest.java [hunk 27]
```diff
@@ -0,0 +1,467 @@
...
+                }
+            } catch (InterruptedException ignored) {
+            } finally {
+                doneLatch.countDown();
+            }
+        }).start();
+
+        startLatch.countDown();
+        assertTrue("Concurrent reset+query should complete within 30s",
+                doneLatch.await(30, java.util.concurrent.TimeUnit.SECONDS));
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java]: Fixed — refactored to ExecutorService with shutdownNow() in finally block (92afb9b).

File: src/main/java/ai/spice/SpiceClient.java [hunk 28]
```diff
@@ -215,6 +223,26 @@ public SpiceClient(String appId, String apiKey, URI flightAddress, URI httpAddre
                 : memoryLimitMB * BYTES_PER_MB;
         this.allocator = new RootAllocator(memoryLimitBytes);
 
+        // Build the Flight client (channel + auth handshake)
+        buildFlightClient();
+
+        // Initialize cached retryers (immutable, built once)
+        initRetryers();
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/main/java/ai/spice/SpiceClient.java]: Fixed — wrapped buildFlightClient/initRetryers in try-catch, closes allocator on failure (92afb9b).

File: src/test/java/ai/spice/ResetTest.java [hunk 29]
```diff
@@ -0,0 +1,467 @@
...
+                assertEquals("Second query after reset should return 2 rows", 2, rows2);
+            }
+        } catch (Exception e) {
+            // Skip if no local runtime or TPC-H data available
+            String msg = e.getMessage() != null ? e.getMessage().toLowerCase() : "";
+            if (msg.contains("unavailable") || msg.contains("connection refused")
+                    || msg.contains("not found") || msg.contains("io exception")) {
+                return;
+            }
+            fail("Unexpected error: " + e.getMessage());
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java]: Fixed — integration tests now use setUp() availability probe with boolean flag, matching TpchIntegrationTest pattern (92afb9b).

File: src/test/java/ai/spice/ResetTest.java [hunk 30]
```diff
@@ -0,0 +1,467 @@
...
+                }
+                assertTrue("Second parameterized query after reset should return rows", rows2 > 0);
+            }
+        } catch (Exception e) {
+            String msg = e.getMessage() != null ? e.getMessage().toLowerCase() : "";
+            if (msg.contains("unavailable") || msg.contains("connection refused")
+                    || msg.contains("not found") || msg.contains("io exception")) {
+                return;
+            }
+            fail("Unexpected error: " + e.getMessage());
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java]: Fixed — same setUp() probe pattern applied to both integration tests (92afb9b).

File: README.md [hunk 31]
```diff
@@ -213,6 +213,41 @@ SpiceClient client = SpiceClient.builder()
...
+
+// Long-lived usage with transport recovery
+try {
+    try (FlightStream stream = client.query(sql)) {
+        // process results...
+    }
+} catch (ExecutionException e) {
+    if (isTransportFailure(e.getCause())) {
+        client.reset();                     // discard bad transport, reconnect immediately
+        try (FlightStream stream = client.query(sql)) {
```
[the remark this answers is not in the document]
Comment (lukekim) [reply README.md]: Fixed — added comment explaining isTransportFailure() is application-defined with suggested exception types (92afb9b).

File: src/test/java/ai/spice/ResetTest.java [hunk 32]
```diff
@@ -0,0 +1,467 @@
...
+                    errors.incrementAndGet();
+                } finally {
+                    doneLatch.countDown();
+                }
+            }).start();
+        }
+
+        startLatch.countDown(); // release all threads
+        assertTrue("Concurrent reset should complete within 30s",
+                doneLatch.await(30, java.util.concurrent.TimeUnit.SECONDS));
```
[the remark this answers is not in the document]
Comment (lukekim) [reply src/test/java/ai/spice/ResetTest.java]: Fixed — both concurrent tests refactored to ExecutorService with proper shutdown/shutdownNow (92afb9b).
